ships/0022-build-strategy-volumes.md

```diff
@@ -0,0 +1,179 @@
+
+- Securing volume usage/access with admission control.
+- Allow developers to inject arbitrary volumes into `BuildRun`s.
+
+## Proposal
+
+### User Stories
+
+#### Share information in build steps
+
+As a build strategy author
+I want to declare volumes in my build steps
+So that I can share information across build steps
+
+#### Use persistent volumes in build runs
+
+As a developer
+I want to declare persistent volume claims in my Builds
+So that I can cache resources like image layers and artifacts for subsequent BuildRuns
+
+### Implementation Notes
+
+The `BuildStrategy` API will be enhanced so that volumes can be declared alongside the build strategy steps.
+The volumes will provide fixed mount points that will be used in all steps, along with a default source for the underlying filesystem.
+Volumes can be marked `optional` indicating that the volume is not needed for a `BuildRun` to execute.
+Volumes can also indicate if the source can be overrode by a `Build` or `BuildRun`.
+
+Both the `Build` and `BuildRun` APIs will be enhanced to allow volumes to be specified.
+The volume name referenced in the `Build` or `BuildRun` must be declared in the build strategy.
+
+#### Sample API
+
+This is a sample of the API specified above.
+Field names and behavior are subject to change once this proposal is deemed implementable.
+
+`BuildStrategy`:
+
+```yaml
+metadata:
+  name: cached-docker-build
+spec:
+  buildSteps:
+  - name: build
+  ...
+  volumes:
+  - name: build-metadata
+    description: "Build metadata"
+    mountPath: /home/build/metadata
+    volumeSource:
+      emptyDir: {}
+  - name: image-cache
+    description: "Container image cache"
+    optional: true
+    mountPath: /var/lib/containers/cache
+    volumeSource:
+      emptyDir: {}
+    allowSourceOverride: true
+```
+
+`Build` and `BuildRun`:
+
+```yaml
+spec:
+  ...
+  strategy:
+    name: cached-docker-build
+    kind: BuildStrategy
+  volumes:
+  - name: image-cache
+    volumeSource:
+      persistentVolumeClaim:
+        name: pvc-image-cache
+```
+
+### Test Plan
+
+TBD
+
+### Release Criteria
+
+#### Removing a deprecated feature [if necessary]
+
+Not applicable
+
+#### Upgrade Strategy [if necessary]
+
+TBD
+
+### Risks and Mitigations
+
+Security is a concern with volumes, especially if arbitrary `HostPath` volume mounts are allowed in the API.
+This can be mitigated through admission controllers, which as a last resort can gate pod admission.
+Ideally we craft documentation which helps admins block admission at `Build` or `BuildRun` creation.
+
+## Drawbacks
+
+Volumes support adds complexity to the API that may require developers to fully inspect the referenced build strategies.
+We have assumed with other proposals that developers should not be expected to interact fully with `BuildStrategy` objects.
+
+
+## Alternatives
+
+We could implement similar capabilities by exposing Tekton's [Workspaces](https://tekton.dev/docs/pipelines/workspaces/) feature set directly.
```

Comment:
Even when writing Tasks outside of the context of Shipwright, I usually used volumes instead of workspace. The reason is that certain volumes, especially if those are secrets, I only wanted to be mounted at the step that needed it rather than all. Now that Tekton recently introduced [isolated workspaces](https://github.com/tektoncd/pipeline/blob/main/docs/workspaces.md#isolated-workspaces) as an alpha feature, I could revisit my own decision. But, no matter which route we take, I definitely want that a BuildStrategy author can define which steps a volume/workspace is mounted to.

Reply:
> I definitely want that a BuildStrategy author can define which steps a volume/workspace is mounted to.

I initially wanted to keep things simple, taking a cue from Tekton Workspaces and have the volume mount in all build steps. Perhaps I should step back and use the more traditional volume/volume mounts approach?

Reply:
Yes, as commented above.